src/events/google_cloud.js

```diff
@@ -79,6 +86,135 @@ function bigQueryWrapper(wrappedFunction) {
+                        callback(err, arg2, ...arg3);
+                    }
+                };
+            });
+            tracer.addEvent(pubsubEvent, responsePromise);
+        } catch (err) {
+            tracer.addException(err);
+        }
+        return original.apply(this, [config, patchedCallback]);
+    };
+}
+
+/**
+ * Wrap pubsub request function.
+ * @param {Function} original pubsub request function.
+ * @returns {Function} The wrapped function
+ */
+function wrapPubSubPullFunction(original) {
+    return function internalPubSubPullFunction(request, options, callback) {
+        try {
+            let pubsubProjectId;
+            if (request && request.subscription) {
+                const subscriptionSplited = request.subscription.split('/');
+                if (subscriptionSplited.length > 1) {
+                    [, pubsubProjectId] = subscriptionSplited;
+                }
+            }
+            const { slsEvent: pubsubEvent, startTime } = eventInterface.initializeEvent(
+                GOOGLE_CLOUD_TYPES.pubsub.name, pubsubProjectId, 'Pull'
+            );
+            const patchedCallback = (err, res) => {
+                const callbackResponse = {};
+                if (res && res.receivedMessages) {
+                    const receivedMessages = res.receivedMessages.reduce((acc, current) => {
+                        acc.push({ messageId: current.message.messageId, message: `${current.message.data}` });
+                        return acc;
+                    }, []);
+                    callbackResponse.receivedMessages = receivedMessages;
+                }
+                eventInterface.finalizeEvent(pubsubEvent, startTime, err, callbackResponse);
+                if (callback) {
+                    callback(err, res);
+                }
+            };
+            // in case callback has given from the client.
+            if (callback) {
+                let patchedCallbackWithPromise = callback;
+                const promise = new Promise((resolve) => {
+                    patchedCallbackWithPromise = (err, res) => {
+                        patchedCallback(err, res);
+                        resolve();
+                    };
+                });
+                tracer.addEvent(pubsubEvent, promise);
+                return original.apply(this, [request, options, patchedCallbackWithPromise]);
+            }
+            const responsePromise = original.apply(this, [request, options, callback]);
+            tracer.addEvent(pubsubEvent, responsePromise);
+            return responsePromise.then((res) => {
+                const [response] = res;
+                patchedCallback(null, response);
+                return res;
+            }, (err) => {
+                patchedCallback(err, null);
+                throw err;
+            });
+        } catch (err) {
+            tracer.addException(err);
+        }
+        return original.apply(this, [request, options, callback]);
```

Comment:
this flow is still buggy:
1. user reaches line 201
2. user reaches line 202 - exception is raised
3. user reaches line 212
4. user reaches line 214

you called the original twice

Reply:
tnx! fixed.